arroyo/types.py

```diff
@@ -85,6 +89,66 @@ def __setstate__(self, state: dict[str, int | datetime]) -> None:
             object.__setattr__(self, slot, value)
 
 
+@dataclass(unsafe_hash=True)
+class BatchPayload(Payload[TPayload]):
```

Comment:
I like the idea that a `Payload` carries a `committable` property that can be managed by the logic that performs the commit in an abstract way.

Though I am not fully getting the naming of the BatchPayload and BrokerPayload. I am not sure I understand the difference of the two concepts and I think we may want to be a bit more opinionated. 

The way this is written, `BrokerPayload` is semantically a subset of a `BatchPayload`. Which means I could just use `BatchPayload` everywhere even if the `__payload` attribute represented one message only. The only issue would be with the `teimstamp` which is not there on the batch payload. 
This is a bit confusing as it make me ask myself why does `BrokerPayload` even exists.

I think we should understand which scenario we want to allow. I can think of:

1. A single message from the broker 
2. A single message that derives from a message coming from the broker (same position but different payload)
3. A batch of messages from the broker 
4. A batch of derived messages 
5. An empty batch with some offset to commit ?
6. A message that is not tied in any way to a message coming from the broker (thus no offset and partition) ?

I think 1 and 2 can easily managed by the same data structure. That's what we do today.
3 and 4 can be managed by the same data structure.
6 Is an interesting question. Should we support anything like that ? I think we could be opinionated in saying no to that. All messages we pass around are related to something to be committed. Can you think about a legit use case for this? Potentially a new message we want to produce somewhere, but why would we pass it through strategies before? If we do not allow this scenario, we are precluding this use case. If we allow that we can allow for an empty committable dictionary.

5 Is another interesting case. I can think about a scenario where we want to propagate am offset to commit even if we filtered out all the messages but I am not sure the Payload abstraction is the right one or whether there should be a special case of Payload without payload for this scenario.

Moreover why not making the `BatchPayload` opinionated in holding a Sequence of `TPayload` ? 
If we do this, Is there any reason why we need a `BrokerPayload` and not having a `BatchPayload` with one element? Is it to make strategies simpler? Do you have a specific use case in mind.

Reply:
>  I think we do not want to enforce that it must be an Array of TPayload. It can equally represent a filtered message with no data but only offsets to commit.

Then why not make it crystal clear to the user by having three types of payload:
- OneMessage
- Batch (that is an array)
- EmptyMessage (that only contains offsets without payload). 

So we make it clear what should be used for what just by reading the name.

Reply:
EmptyMessage is something that we may want in the future. But I think we cannot support easily support it today without fairly significant changes. I think the biggest challenge is that it means that every strategy effectively needs to know about EmptyMessage and handle it appropriately, including user defined ones. I think there are still too many open questions for this to be included now.

About the Batch as an array, I think there are other reasons too why we shouldn't do it:
- If you look at the batch implementation in https://github.com/getsentry/arroyo/pull/127/ the `MessageBatch` object is not an array and contains other useful metadata about the batch.
- From a typing perspective, I think it will also make life more complicated. All messages passed to any strategy conform to `Message[TPayload]` and the typechecker ensures that `TOutput` of one strategy the `Message[TPayload]` of the next strategy. If we force BatchPayload to be `Message[List[TPayload]]` then I don't see how our common message class will work anymore as the type would not line up
- In the end I think the user can already choose to make TPayload = List[T] and forcing the type into an array doesn't make their life any meaningfully better

Reply:
> EmptyMessage is something that we may want in the future. But I think we cannot support easily support it today without fairly significant changes. I think the biggest challenge is that it means that every strategy effectively needs to know about EmptyMessage and handle it appropriately, including user defined ones

That's fair, though how do you envision the problem (passing offset to commit) being managed in the implementation after this PR? Do we expect strategies that perform commit on behalf of others to be able to interpret an empty batch as some offsets to commit without any content ?

Reply:
At this time, if we want to have the ability to:
- filter stuff at the start of the pipeline (or anywhere)
- all strategies can still work with all others
- users can easily extend and build their own strategies but they must handle the empty message too
- actually commit at least some those filtered messages

... then the only solution I can think of involves every single strategy knowing about some new `EmptyMessage` type or value and handling it in a custom way. In many cases it's probably just forwarding to the next step.

In this scenario all we can do is try and make this as un-painful for users as possible but it does add at least a little complexity to every strategy.

Thoughts?

Reply:
> If we want to have the ability to:

I believe this should be considered a requirement. At least this part: "A strategy needs to be able to filter messages at any point (basic product requirements). Periodically filtered out offsets must be committed even if there is no fully processed message (otherwise you accumulate messages in the Kafka queue and you may end up hitting Kafka retention breaking the consumer). An empty message cannot be committed until all the previous messages have been fully processed (basic consistency).

I think your idea `EmptyMessage`  is a valid solution.

Now, do we even have any alternative?
- In order to make filtering hidden from the steps that follow the filtering one, the message would have to be discarded early in the pipeline
- In order for offset to be still committed in the right order, we cannot commit in the place where we discard 
- This means that whoever commits needs to have visibility of the filtered messages.

This implies that, either the filtered message go through the pipeline anyway in some way or the processor keeps track of all messages and commits on its own when a number of filtered message is accumulated without any real message being present. 
Option two would basically go against the idea of keeping the commit process visible to the strategy.

So I don't think we really have an alternative that does not entirely hide the commit process from the application developer.

Reply:
No I don't think there's any alternatives, there might be a few options for how `EmptyMessage` gets implemented but the outcome is the same. Every processing step just needs to handle it and pass it along to the next. Anything else is wrong from a consistency perspective.

Reply:
I added an issue for this here https://github.com/getsentry/arroyo/issues/143. Let's move the discussion there. I think that this new message interface might might potentially make this a little easier (we can send one filtered message across many partitions instead of one per partition), but the problem is mostly unrelated to this change.